### PR TITLE
Add broker truncation via aging.

### PR DESCRIPTION
**Pull request by benbjohnson**
## Overview

This pull request updates the broker truncation to drop segments when they become too old. We should also include the ability to actively delete topics on the broker but communicating between the data node and broker can fail so we should still have a fallback to eventually delete segments.

Partially fixes: https://github.com/influxdb/influxdb/issues/2382

### TODO

- [ ] Add additional logging.
- [ ] Add replication check for stale segments.

### CODE REVIEW

File: messaging/broker.go [hunk 1]
```diff
@@ -965,30 +977,34 @@ func (t *Topic) Truncate(maxSize int64) (int, int64, error) {
...
-			break
-		}
-
-		// Attempt deletion of oldest segment in topic. First and second segment needed
-		// for this operation.
-		first := segments[0]
-		second := segments[1]
+		// A segment can be deleted under two conditions:
+		//
+		//   1) the segment is too stale
```
Comment (otoolep) [inline messaging/broker.go]: This is an AND, right? You might like to make this clear in the comments.

File: messaging/broker.go [hunk 2]
```diff
@@ -965,30 +977,34 @@ func (t *Topic) Truncate(maxSize int64) (int, int64, error) {
...
 		}
 
-		// More than 2 segments are needed for current writes and replication checks.
-		if len(segments) <= 2 {
-			break
-		}
-
-		// Attempt deletion of oldest segment in topic. First and second segment needed
-		// for this operation.
-		first := segments[0]
```
Comment (otoolep) [inline messaging/broker.go]: You don't like `first` and `second`? :-)

File: messaging/broker.go (hunk 1 above)
Comment (otoolep) [inline messaging/broker.go]: Oh, I see it's an OR. Are we sure about this?

**Comment by otoolep** [comment]
I want to be clear on the policy. This is going to mean data is deleted once it gets too old, regardless of whether it's been replicated.

File: messaging/broker.go [hunk 3]
```diff
@@ -965,30 +977,34 @@ func (t *Topic) Truncate(maxSize int64) (int, int64, error) {
...
-		second := segments[1]
+		// A segment can be deleted under two conditions:
+		//
+		//   1) the segment is too stale
+		//   2) the last entry has been replicated
+		//
+		// The most efficient way to see if the last entry has been replicated
+		// is to check if first entry of next segment has been replicated.
+		if len(segments) > 1 && t.MaxSegmentAge > 0 && time.Since(segments[0].ModTime) > t.MaxSegmentAge {
```
Comment (jwilder) [inline messaging/broker.go]: Just to clarify, if data node subscribers have not pulled down this data in two days, it would get truncated and removed?

Comment (otoolep) [inline messaging/broker.go]: Yeah, exactly. We do need to policy like this, or data would grow forever on the brokers. I too want to be sure about introducing it now.

Comment (otoolep) [inline messaging/broker.go]: In the event of a permanently downed data node that is.

**Comment by otoolep** [comment]
I feel we need logging added to this change. The reasons for data truncation are now becoming varied, and operators will need to know *why* data is getting deleted on their brokers.

**Comment by benbjohnson** [comment]
@otoolep @jwilder I can add a check to see if it's replicated but there's two issues:

1. Replication data isn't persisted on the brokers so if a shard is deleted and the brokers are restarted then data nodes will not heartbeat any longer. This will cause the segments to remain indefinitely.

2. Once we get down to 1 segment we'll need to read the whole thing in to determine the last index. That's expensive although it's probably not a huge issue since it shouldn't happen often.

**Comment by benbjohnson** [comment]
I agree about the logging.

**Comment by jwilder** [comment]
Seems like when a shard is deleted by a data node, it should notify the broker that the that topic can be removed.  I also think it would be less surprising to clients if new writes started to fail (because topic is full) as opposed to writes that were acknowledged and then disappeared two days later unexpectedly.

File: messaging/broker.go [hunk 4]
```diff
@@ -33,6 +33,9 @@ const DefaultMaxTopicSize = 1024 * 1024 * 1024 // 1GB
 // DefaultMaxSegmentSize is the largest a segment can get before starting a new segment.
 const DefaultMaxSegmentSize = 10 * 1024 * 1024 // 10MB
 
+// DefaultMaxSegmentAge is the longest a stale segment can exist before removal.
+const DefaultMaxSegmentAge = 2 * 24 * time.Hour // 2 days
```
Comment (jwilder) [inline messaging/broker.go]: Should make this configurable via the config file too.

**Comment by benbjohnson** [comment]
@jwilder I agree that data nodes should notify the broker on deletion but if the message to the broker  fails then how does it recover? It's in the broadcast topic so we can't block until we reach the broker. We'd have to have a separate goroutine to monitor the state between the brokers & the data nodes.

**Pull request closed without merging**